### PR TITLE
Priyanka

### PR DESCRIPTION
**Pull request by Priya411**

### CODE REVIEW

File: src/codeu/chat/common/ConversationHeader.java [hunk 1]
```diff
@@ -63,4 +63,21 @@ public ConversationHeader(Uuid id, Uuid owner, Time creation, String title) {
     this.title = title;
 
   }
+
+  @Override
+  public boolean equals(Object toCompare)
+  {
+    if(!(toCompare instanceof ConversationHeader))
```
Comment (KevinWorkman) [inline src/codeu/chat/common/ConversationHeader.java]: nit: imho, you should always use `{}` curly braces after an `if` statement, even if it only contains a single statement.

Here is a link to Google's style guide talking about this: https://google.github.io/styleguide/javaguide.html#s4.1-braces

But it's not just a Google thing. Here is a link to a Stack Overflow discussion: https://stackoverflow.com/questions/8020228/is-it-ok-if-i-omit-curly-braces-in-java

Basically, omitting the curly braces makes it more likely that somebody will introduce code like this:

```
if(!(toCompare instanceof ConversationHeader))
   System.out.println("not an instance of ConversationHeader");
   return false;
```

Since there aren't curly brackets, this code will always return false.

File: src/codeu/chat/common/ConversationHeader.java [hunk 2]
```diff
@@ -63,4 +63,21 @@ public ConversationHeader(Uuid id, Uuid owner, Time creation, String title) {
...
+      return false;
+    ConversationHeader toCompareConv = (ConversationHeader)(toCompare);
+    if(!this.id.equals(toCompareConv.id))
```
Comment (KevinWorkman) [inline src/codeu/chat/common/ConversationHeader.java]: Is it possible for `this.id` to be `null`? Similar question for all of the `if` statements in this PR.

File: src/codeu/chat/common/ConversationHeader.java [hunk 3]
```diff
@@ -63,4 +63,21 @@ public ConversationHeader(Uuid id, Uuid owner, Time creation, String title) {
     this.title = title;
 
   }
+
+  @Override
+  public boolean equals(Object toCompare)
```
Comment (KevinWorkman) [inline src/codeu/chat/common/ConversationHeader.java:69]: Remember that if you override `equals()`, you should also override `hashcode()`. More info here: https://stackoverflow.com/questions/2265503/why-do-i-need-to-override-the-equals-and-hashcode-methods-in-java

Basically, if you'll ever use these objects in a `HashMap` or a `HashSet`, they'll first use `hashcode()` to find items, and then they'll use `equals()` to compare them. You don't want two items to be equal (according to your new `equals()` function) but **not** equal according to `hashcode()`.

File: src/codeu/chat/server/JSON.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            // If the file exists, then the file will be parsed
+           // in this set of code
+           jp.nextToken();
+            while (jp.nextToken() != JsonToken.END_OBJECT)
+            // This ensures that we haven't reached the end of the
+            // file because the data is being parsed in such a way
+            // that this END_OBJECT specifically relates to the
+            // very last } in the json file
+            {
+            if(jp.getText()=="users") {
```
Comment (KevinWorkman) [inline src/codeu/chat/server/JSON.java]: You should almost never use `==` to compare `String` values. Use the `equals()` function instead. More info here: https://stackoverflow.com/questions/513832/how-do-i-compare-strings-in-java

But basically, `==` compares whether two values are the same instance. This might work because of the way `String` values are stored behind the scenes, but it's almost definitely not what you want.

Make sure to fix this in other places where you're using `==` with `String` values too.

File: src/codeu/chat/server/JSON.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                // If the object type represents a User, then
+                // specific steps will take place to create
+                // a User variable
+                jp.nextToken();
+                while (jp.nextToken() != JsonToken.END_ARRAY) {
+                    String name = "";
+                    Uuid id = null;
+                    Time time = null;
+                    jp.nextToken();
+                    if (jp.getText() == "name") {
```
Comment (KevinWorkman) [inline src/codeu/chat/server/JSON.java]: Does this mean that the order of the fields in your JSON matters?

File: src/codeu/chat/util/Time.java [hunk 6]
```diff
@@ -53,6 +53,13 @@ public int compareTo(Time other) {
 		return date.compareTo(other.date);
 	}
 
+	@Override
+	public boolean equals (Object toCompare)
+	{
+		if(!(toCompare instanceof Time))
+			return false;
+		return this.toString().equals(((Time)(toCompare)).toString());
```
Comment (KevinWorkman) [inline src/codeu/chat/util/Time.java]: I don't think you need to case to `Time` here?

File: test/codeu/chat/server/JSONTest.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+import java.util.Iterator;
+import codeu.chat.util.Time;
+
+/**
+ * This class tests the JSON class in the server folder
+ * The purpose of this test is to ensure that the user is able to
+ * access the transaction log and that the transaction log is
+ * being saved properly.
+ */
+public class JSONTest {
```
Comment (KevinWorkman) [inline test/codeu/chat/server/JSONTest.java:24]: Thank you for writing tests! Nice work. :tada:

File: test/codeu/chat/server/JSONTest.java [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+    @Test
+    public void readTest1() {
```
Comment (KevinWorkman) [inline test/codeu/chat/server/JSONTest.java]: I would rename your test functions to be more obvious. Something like `testReadAll` or `testReadUser`, names like that.

File: test/codeu/chat/server/JSONTest.java [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+            // Path used by Priyanka:
+            model = log.readFromFile("/Users/HMCLoaner/Desktop/CodeU/test/codeu/chat/server/input.json");
+            // The path can vary from device, so this may not compile if it's not the correct path for your project
+            // set up. A suggested path is as follows
+            // model = log.readFromFile("CodeU/test/codeu/chat/server/input.json");
+        } catch (IOException e) {
+            System.out.println("File not handled properly");
+        }
+        assertEquals(
+                compare(model.userById().all().iterator(), comparer.userById().all().iterator()), true);
```
Comment (KevinWorkman) [inline test/codeu/chat/server/JSONTest.java]: Is there a way to just compare the collections directly instead of going to the iterator? What does `model.userById()` return?

Comment (KevinWorkman) [reply test/codeu/chat/server/JSONTest.java]: Like, can't you do something like `assertEquals(model.userbyId(), comparer.userById());`?

File: test/codeu/chat/server/JSONTest.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+ * being saved properly.
+ */
+public class JSONTest {
+
+    @Test
+    public void readTest1() {
+        // This function tests if the JSON reader is working if there is a JSON with one type of each Object in the file
+        Model model = new Model();
+        JSON log = new JSON();
+        Model comparer = new Model();
```
Comment (KevinWorkman) [inline test/codeu/chat/server/JSONTest.java]: nit: I would maybe call this `expected` instead of `comparer`?

File: test/codeu/chat/server/JSONTest.java [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+    @Test
+    public void readTest2() {
+        // This function tests if the JSON reader is working if there is a JSON with one type of User in the file
+        Model model = new Model();
+        JSON log = new JSON();
+        Model comparer = new Model();
+        comparer.add(new User(createUuid("1.1074501217"), "Krager", Time.fromMs(12345)));
+        try {
+            // Path used by Priyanka:
+            model = log.readFromFile("/Users/HMCLoaner/Desktop/CodeU/test/codeu/chat/server/input2.json");
```
Comment (KevinWorkman) [inline test/codeu/chat/server/JSONTest.java]: Could you include this in the `src` directory and use a relative path?

File: test/codeu/chat/server/JSONTest.java [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+        try {
+            // Path used by Priyanka:
+            model = log.readFromFile("/Users/HMCLoaner/Desktop/CodeU/test/codeu/chat/server/input3.json");
+            // The path can vary from device, so this may not compile if it's not the correct path for your project
+            // set up. A suggested path is as follows
+            //model = log.readFromFile("CodeU/test/codeu/chat/server/input3.json");
+        } catch (IOException e) {
+            System.out.println("File not handled properly");
+        }
+        assertEquals(
```
Comment (KevinWorkman) [inline test/codeu/chat/server/JSONTest.java]: If all of these `assertEquals()` statements are the same in every function, you might extract them into a helper function that takes the `model` and `comparer` as parameters?

File: test/codeu/chat/server/JSONTest.java [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+                compare(model.conversationByText().all().iterator(), comparer.conversationByText().all().iterator()), true);
+        assertEquals(
+                compare(model.messageById().all().iterator(), comparer.messageById().all().iterator()), true);
+        assertEquals(
+                compare(model.messageByText().all().iterator(), comparer.messageByText().all().iterator()), true);
+        assertEquals(
+                compare(model.messageByTime().all().iterator(), comparer.messageByTime().all().iterator()), true);
+    }
+
```
Comment (KevinWorkman) [inline test/codeu/chat/server/JSONTest.java]: I would maybe expect to see tests for bad cases as well. What happens if the JSON contains invalid data? Or other corner cases? Tests should make sure your code works in corner cases, not just in cases you expect.

**Review by KevinWorkman — COMMENTED** [review]
Overall looks good, awesome work! Few comments below.

File: src/codeu/chat/common/ConversationHeader.java (hunk 2 above)
Comment (Priya411) [reply src/codeu/chat/common/ConversationHeader.java]: I believe so. If it isn't there in the JSON file at all, then the default value is null and it works. In addition, there are some values in Message which I'm setting to null since we just don't have that data and it is working.

File: test/codeu/chat/server/JSONTest.java (hunk 9 above)
Comment (Priya411) [reply test/codeu/chat/server/JSONTest.java]: That returns a type StoreAccessor. I couldn't really find a clear way to implement a equals function or any other method of comparison until I got down to the iterator.

File: test/codeu/chat/server/JSONTest.java (hunk 11 above)
Comment (Priya411) [reply test/codeu/chat/server/JSONTest.java]: I'd originally put it in the src file but moved it because I thought it may be better for the files to be together. I don't know why, but it won't allow me to use a relative path. It isn't getting the file.

File: src/codeu/chat/common/ConversationHeader.java (hunk 1 above)
Comment (Priya411) [reply src/codeu/chat/common/ConversationHeader.java]: I added curly bracket for all of the functions I made, which were in ConversationHeader, Message, and Time

File: test/codeu/chat/server/JSONTest.java (hunk 8 above)
Comment (Priya411) [reply test/codeu/chat/server/JSONTest.java]: I renamed them with more clear titles

File: test/codeu/chat/server/JSONTest.java (hunk 10 above)
Comment (Priya411) [reply test/codeu/chat/server/JSONTest.java]: I changed all instances of comparer into expected

File: test/codeu/chat/server/JSONTest.java (hunk 12 above)
Comment (Priya411) [reply test/codeu/chat/server/JSONTest.java]: I made a new function with these assertEquals statements in them.

File: src/codeu/chat/server/JSON.java (hunk 5 above)
Comment (Priya411) [reply src/codeu/chat/server/JSON.java]: Yes, the fields matter but the order of object types doesn't

File: src/codeu/chat/server/JSON.java (hunk 4 above)
Comment (Priya411) [reply src/codeu/chat/server/JSON.java]: I fixed this all the times I did it

File: src/codeu/chat/util/Time.java (hunk 6 above)
Comment (Priya411) [reply src/codeu/chat/util/Time.java]: I tested it, and I actually didn't, so I removed it.

File: test/codeu/chat/server/JSONTest.java (hunk 13 above)
Comment (Priya411) [reply test/codeu/chat/server/JSONTest.java]: I added more test

File: src/codeu/chat/common/ConversationHeader.java (hunk 3 above)
Comment (Priya411) [reply src/codeu/chat/common/ConversationHeader.java:69]: If the ConversationHeader class and Messages class are ultimately using the equals() of Uuid and Time only, should the hashcode function still be overwritten? Or should I call the haschode functions of those?

Comment (Priya411) [reply src/codeu/chat/common/ConversationHeader.java:69]: Actually, because of research, I added the hashcode() for everything I overwrote equals() for

File: test/codeu/chat/server/JSONTest.java (hunk 11 above)
Comment (KevinWorkman) [reply test/codeu/chat/server/JSONTest.java]: I think relative paths are relative to the directory the code was run from, not the location of the class file.  You might also have to modify the `build.py` file to copy over `.json` files into the `bin` directory. Because when the code is executing, that's where it's executing from. Basically it has to be relative to the `.class` file, not the `.java` file.

**Review by KevinWorkman — APPROVED** [review]
LGTM

I think the only remaining potential issue was whether the variables inside some of these classes can be `null` or not. If not you're okay, but if so you'll probably need to add null checks. That problem will be pretty obvious though, so maybe it's okay to wait to see how it works.

I'd say check with your teammates, but from my perspective this code is ready to be merged to master.
